include/fmt/ranges.h

```diff
@@ -390,22 +390,36 @@ using range_formatter_type = conditional_t<
 template <typename R>
 using maybe_const_range =
     conditional_t<has_const_begin_end<R>::value, const R, R>;
+
+template <typename R>
+struct is_not_recursive_range : bool_constant<
+    !std::is_same<detail::uncvref_type<R>, R>::value> {};
+
+template <typename R, typename Char>
+struct is_formattable_delayed : is_formattable<
+    detail::uncvref_type<detail::maybe_const_range<R>>, Char> {
+};
+
+template <typename R, typename Char>
+struct has_fallback_formatter_delayed : detail::has_fallback_formatter<
+    detail::uncvref_type<detail::maybe_const_range<R>>, Char> {
+};
```

Comment:
What are these for? I'd suggest keeping them inlined as before.

Reply:
The whole point of this P/R is to ensure that the individual parts of the condition are evaluated in short-circuit manner. In C++ without language concepts this is only possible by defining type predicates that get only evaluated if the corresponding predicate template class definition becomes instantiated. Due to `std::conjunction` and friends we therefore need individual class templates. The trailing `delayed` is intended to emphasize their job.

Reply:
Could you clarify this in a comment?

Reply:
I'd suggest either adding a generic thing to delay instantiation, like so:

```cpp
template <template <class...> class F, typename... Args> struct delayed : F<Args...> { };
```
to be used as:
```cpp
disjunction<
    delayed<is_formattable, detail::uncvref_type<detail::maybe_const_range<R>>, Char>,
    delayed<detail::has_fallback_formatter, detail::uncvref_type<detail::maybe_const_range<R>>, Char>
    >
```
or, if we're going to manually do our own thing, do all of it together:
```cpp
template <class T, class Char>
struct delayed_fallback_formattable
    : disjunction<
        is_formattable<T, Char>,
        detail::has_fallback_formatter<T, Char>>
{ };
```
to be used as:
```cpp
conjunction<
    A,
    B,
    delayed_fallback_formattable<detail::uncvref_type<detail::maybe_const_range<R>>, Char>
    >
```

Put differently - instead of writing two bespoke delayed traits, write either one generic delayer or one bespoke delayed trait.

Reply:
This is a good suggestion but note that `has_fallback_formatter` will soon go away so I suggest keeping two bespoke traits. Keeping both will simplify removing `has_fallback_formatter_delayed` later.

Reply:
Oh. Ok that makes sense then. When you say soon, how soon?

Reply:
I merged now has_fallback_formatter_delayed into is_formattable_delayed and got rid of the former.

Reply:
> how soon?

In the next major release about a year from now.